### ISSUE TITLE
Postgresql InvalidColumnReference

### ISSUE THREAD

**Issue by greyman888** [maxrossello/redmine_app_timesheets#56]
Hi Max,

I just set up a new instance of Redmine and used the opportunity to upgrade everything. Here is my new config (also running on Heroku for the first time):

**Environment:**
  Redmine version                3.1.0.stable
  Ruby version                   2.2.1-p85 (2015-02-26) [x86_64-linux]
  Rails version                  4.2.4
  Environment                    production
  Database adapter               PostgreSQL
**Redmine plugins:**
  redmine_app__space             1.0.3
  redmine_app_timesheets         1.5.3

I've set up everything the same as before but when I try to open the timesheet I get an error (log extract below). I'd appreciate any help you can provide.

Thanks,

Adam

2015-09-16T13:08:49.405888+00:00 app[web.1]: Started GET "/apps/timesheets" for 115.42.16.233 at 2015-09-16 13:08:49 +0000
2015-09-16T13:08:49.416717+00:00 app[web.1]:   Current user: adam (id=5)
2015-09-16T13:08:49.438034+00:00 app[web.1]: 
2015-09-16T13:08:49.438038+00:00 app[web.1]: ActiveRecord::StatementInvalid (PG::InvalidColumnReference: ERROR:  for SELECT DISTINCT, ORDER BY expressions must appear in select list
2015-09-16T13:08:49.438040+00:00 app[web.1]: LINE 1: ...s_permissions"."principal_id" IN (5, 6)  ORDER BY access DES...
2015-09-16T13:08:49.438041+00:00 app[web.1]:                                                              ^
2015-09-16T13:08:49.438043+00:00 app[web.1]: : SELECT DISTINCT "ts_permissions"."id" AS t0_r0, "ts_permissions"."order_id" AS t0_r1, "ts_permissions"."access" AS t0_r2, "ts_permissions"."principal_id" AS t0_r3, "ts_permissions"."is_primary" AS t0_r4, "versions"."id" AS t1_r0, "versions"."project_id" AS t1_r1, "versions"."name" AS t1_r2, "versions"."description" AS t1_r3, "versions"."effective_date" AS t1_r4, "versions"."created_on" AS t1_r5, "versions"."updated_on" AS t1_r6, "versions"."wiki_page_title" AS t1_r7, "versions"."status" AS t1_r8, "versions"."sharing" AS t1_r9, "versions"."in_timesheet" AS t1_r10, "versions"."is_order" AS t1_r11 FROM "ts_permissions" INNER JOIN "versions" ON "versions"."id" = "ts_permissions"."order_id" AND "versions"."is_order" = $1 WHERE "ts_permissions"."is_primary" = $2 AND "ts_permissions"."principal_id" IN (5, 6)  ORDER BY access DESC):
2015-09-16T13:08:49.438045+00:00 app[web.1]:   lib/redmine/sudo_mode.rb:63:in `sudo_mode'
2015-09-16T13:08:49.438046+00:00 app[web.1]: 
2015-09-16T13:08:49.438048+00:00 app[web.1]: 
2015-09-16T13:08:49.410011+00:00 app[web.1]: Processing by TimesheetsController#index as HTML
2015-09-16T13:08:49.436402+00:00 app[web.1]: PG::InvalidColumnReference: ERROR:  for SELECT DISTINCT, ORDER BY expressions must appear in select list
2015-09-16T13:08:49.436408+00:00 app[web.1]: LINE 1: ...s_permissions"."principal_id" IN (5, 6)  ORDER BY access DES...
2015-09-16T13:08:49.436410+00:00 app[web.1]:                                                              ^
2015-09-16T13:08:49.436414+00:00 app[web.1]: : SELECT DISTINCT "ts_permissions"."id" AS t0_r0, "ts_permissions"."order_id" AS t0_r1, "ts_permissions"."access" AS t0_r2, "ts_permissions"."principal_id" AS t0_r3, "ts_permissions"."is_primary" AS t0_r4, "versions"."id" AS t1_r0, "versions"."project_id" AS t1_r1, "versions"."name" AS t1_r2, "versions"."description" AS t1_r3, "versions"."effective_date" AS t1_r4, "versions"."created_on" AS t1_r5, "versions"."updated_on" AS t1_r6, "versions"."wiki_page_title" AS t1_r7, "versions"."status" AS t1_r8, "versions"."sharing" AS t1_r9, "versions"."in_timesheet" AS t1_r10, "versions"."is_order" AS t1_r11 FROM "ts_permissions" INNER JOIN "versions" ON "versions"."id" = "ts_permissions"."order_id" AND "versions"."is_order" = $1 WHERE "ts_permissions"."is_primary" = $2 AND "ts_permissions"."principal_id" IN (5, 6)  ORDER BY access DESC
2015-09-16T13:08:49.436846+00:00 app[web.1]: Completed 500 Internal Server Error in 27ms (ActiveRecord: 8.6ms)
2015-09-16T13:08:49.441710+00:00 heroku[router]: at=info method=GET path="/apps/timesheets" host=redminex.herokuapp.com request_id=5659ef96-53a4-4dc6-b308-09d2afa17e6a fwd="115.42.16.233" dyno=web.1 connect=1ms service=34ms status=500 bytes=839

**Comment by maxrossello** [comment]
Hi Adam,
the issue may be solvable through the following advice:
http://stackoverflow.com/questions/9758793/rails-3-activerecord-postgresql-uniq-command-doesnt-work/11476714#11476714

The problem is that unfortunately I am not succeeding in identifying the offending line in code given this kind of log...

**Comment by greyman888** [comment]
Thanks for the advice Max, I think it's in one (or all) the following lines in the time_sheets controller:
```
  def get_timelogs
    @active_orders = (TsPermission.for_user(@user).joins(:order).includes(:order).map{|t| t.order}.compact +
        WorkOrder.visible(@user)).uniq.sort_by{ |v| v.name.downcase}
    @active_own_orders = (TsPermission.for_user(User.current).joins(:order).includes(:order).map{|t| t.order}.compact +
        WorkOrder.visible(User.current)).uniq.sort_by{ |v| v.name.downcase}
    # native orders assigned to user + visible versions marked as orders
    # + orders associated to existing timelogs even if version no more visible to user
    # + orders associated to issues that are associated to some existing user timelog even if version no more visible to user
    @orders = (@active_orders +
        WorkOrder.eager_load(:time_entries).merge(TsTimeEntry.for_user(@user)).uniq.all +
        WorkOrder.joins(:issues).merge(Issue.joins(:time_entries).where('time_entries.user_id = ?', @user.id)).uniq.all
    ).uniq.sort_by{ |v| v.name.downcase}
```

At the moment I don't have the time to try and figure out what needs to be changed so I'm going to swap from postgresql to mysql and see how that goes.

Close this issue if you want to and if I need to go back to postgresql later I'll open another issue and try to make the fix then.

Adam

**Comment by greyman888** [comment]
Max,
I think I’ve fixed the problem and I’ve submitted a pull request #63 63. I’ve tested the changes using mysql and postgres and they both seem to work. The change I’ve made only affects native orders so that is the focus of my testing.

Here are the test configurations:
**Postgresql testing**
  Redmine version: 3.1.1.devel
  Ruby version: 2.2.3-p173 (2015-08-18) [x86_64-linux]
  Rails version: 4.2.4
  Environment: development
  Database adapter: PostgreSQL
  redmine_app__space: 1.0.4
  redmine_app_timesheets: 1.5.4 (with my modification to timesheets_controller.rb)
**Mysql testing**
  Redmine version: 3.1.0.stable
  Ruby version: 2.2.1-p85 (2015-02-26) [x86_64-linux]
  Rails version: 4.2.4
  Environment: development
  Database adapter: Mysql2
  redmine_app__space: 1.0.3
  redmine_app_timesheets: 1.5.3 (with my modification to timesheets_controller.rb)

**Here is the test set up:**
I created 1 user who did not have access to the orders application
I created 4 native orders:
1.	Order 1 called “User has access”
2.	Order 2 called “User does not have access”
3.	Order 3 called “User has access but it is later removed”
4.	Order 4 called “User has access but order is later disabled”

Test 1.
Setup: Administrator adds the user to Orders 1,3,4.
User Actions: User opens timesheet, adds all the visible orders to his timesheet and adds time to each of them.
Expected Result: Order 2 should not be visible in the dropdown box. Order 1,3,4 should be visible.

Test 2.
Setup: The user is removed from Order 3 by the administrator
User Actions: User opens timesheet and adds time to all visible orders.
Expected Results: Order 3 should not be visible in the dropdown box & user should not be able to add more time to this order in their time sheet. Row should be coloured grey.

Test 3. 
Setup: Order 4 is disabled by the administrator
User Actions: User opens timesheet and adds time to all visible orders.
Expected Results: Order 4 should not be visible in the dropdown box & user should not be able to add more time to this order in their time sheet. Row should be coloured grey.

**Test Results**
In the mysql testing, all tests passed.
In the postgresql testing, Test 1 and Test 3 passed but Test 2 only partially passed. In this test, Order 3 was not visible in the dropdown box as expected, but the user was able to add more time to the order and the order row was not grey.

I believe that this could be the result of a separate bug and not related to the change that I’ve made. This bug is much less severe than the existing bug and I think we can live with it for now. The scenario where we remove people’s access to codes is very infrequent (I don’t think we have ever done it). We generally enable/disable codes.

Please let me know if there are any other tests you need me to perform.

**Comment by maxrossello** [comment]
Great job! For those willing to try this I merged the pull request on branch 1.5-dev. I will merge to master after a round of my usual tests.
I wish I had automated them from the beginning... but the project was born as a demo... same recurring mistake :-(

**Comment by maxrossello** [comment]
My tests passed. I am going to merge the code to master.

I will also report the newly found issue on Postgres on a dedicated issue.

Thank you again!
